dao/src/main/scala/za/co/absa/enceladus/dao/rest/CrossHostApiCaller.scala

```diff
@@ -18,43 +18,99 @@ package za.co.absa.enceladus.dao.rest
 import org.apache.commons.lang.exception.ExceptionUtils
 import org.slf4j.LoggerFactory
 import org.springframework.web.client.{ResourceAccessException, RestClientException}
+import za.co.absa.enceladus.dao.rest.CrossHostApiCaller.logger
 import za.co.absa.enceladus.dao.{DaoException, RetryableException}
 
+import scala.annotation.tailrec
 import scala.util.{Failure, Random, Try}
 
 protected object CrossHostApiCaller {
 
-  def apply(apiBaseUrls: List[String]): CrossHostApiCaller = {
-    new CrossHostApiCaller(apiBaseUrls, Random.nextInt(apiBaseUrls.size))
+  private val logger  = LoggerFactory.getLogger(classOf[CrossHostApiCaller])
+
+  private def createInstance(apiBaseUrls: Seq[String],
+                             tryCounts: Seq[Int] = Vector.empty,
+                             startWith: Option[Int] = None): CrossHostApiCaller = {
+    def filterUrls(entry: (String, Int)): Boolean = {
+      entry match {
+        case ("", _)                          => false
+        case (url, tryCount) if tryCount <= 0 =>
+          logger.warn(s"Url $url doesn't have a positive try count ($tryCount). Ignoring.")
+          false
+        case _                                => true
+      }
+    }
+
+    val urls = apiBaseUrls.toVector
+    val tries = tryCounts.toVector
+    val triesExtendedToUrlsLength = tries ++ Vector.fill(urls.size - tries.size)(tries.lastOption.getOrElse(1))
+    val urlsWithTries = (urls zip triesExtendedToUrlsLength).filter(filterUrls)
+    val currentHostIndex = startWith.getOrElse(Random.nextInt(Math.max(urlsWithTries.size, 1)))
+    new CrossHostApiCaller(urlsWithTries, currentHostIndex)
+  }
+
+  def apply(apiBaseUrls: Seq[String]): CrossHostApiCaller = {
+    createInstance(apiBaseUrls)
+  }
+
+  def apply(apiBaseUrls: Seq[String], startWith: Int): CrossHostApiCaller = {
+    createInstance(apiBaseUrls, Vector.empty, Option(startWith))
+  }
+
+  def apply(apiBaseUrls: Seq[String], tryCounts: Seq[Int]): CrossHostApiCaller = {
+    createInstance(apiBaseUrls, tryCounts)
+  }
+
+  def apply(apiBaseUrls: Seq[String], tryCounts: Seq[Int], startWith: Int): CrossHostApiCaller = {
```

Comment:
I am not super happy about this API, but I sort-of understand why you did it.

Most logical would be
```scala
def apply(apiBaseUrls: Seq[String], startWith: Int)
def apply(apiBaseUrls: Seq[(String, Int)], startWith: Int)
```
 -- but this you cannot do because, because of the erasure, these would conflict. But on the other hand, now have to deal with checking the `tryCounts` list to be of matching size and if not, take some measures (extending, capping, ...)

In case we truly need the variance on the number of retries (why do we need this btw? (the config example of `menas.rest.uriTries=[1]` seems to suggest that we don't) and isn't this sort of mimicking the truer load-balancing by giving some URLs more retry-space than the others where the round-robin is not good enough (see comment above)), there may be ways to change the API to both force `tryCount` count to match to URLs and still offer the possibility to omit the `tryCount`, e.g.:

```scala
  case class RetryableUrl(url: String, retryCount: Int) {
    require(retryCount >= 1)
  }

  def apply(apiBaseUrls: Seq[String], startWith: Int): CrossHostApiCaller
  def apply(apiBaseUrls: Seq[RetryableUrl], startWith: Int): CrossHostApiCaller
```

and perhaps support it with some conversion helpers?
```scala
  implicit class RetryableUrlStringExt(val url: String) extends AnyVal {
   def toRetryableUrl(retryCount: Int = 1): RetryableUrl = RetryableUrl(url, retryCount)
  }

  implicit class RetryableUrlStringIntExt(val urlAndRetry: (String, Int)) extends AnyVal {
    def toRetryableUrl: RetryableUrl = RetryableUrl(urlAndRetry._1, urlAndRetry._2)
  }
```